### DISCUSSION TITLE
pip install zxing-cpp - needs git when building from source?

### DISCUSSION THREAD

**Discussion by arechnitzer** [zxing-cpp/zxing-cpp#494] [Q&A] [answered]
I'm pip-installing zxing-cpp in a docker container and trying to keep things as minimal as possible.  

Can one pip-install without first installing git?

**Comment by cbm755** [comment]
note: Ubuntu 20.04 has Python 3.8 and there is no wheel for 3.8 so it builds from source (#488).


fails:
```
docker pull ubuntu:20.04
docker run -it --rm ubuntu:20.04
apt-get update
apt-get install python3-dev python3-pip
pip install zxing-cpp
```

works:
```
docker pull ubuntu:20.04
docker run -it --rm ubuntu:20.04
apt-get update
apt-get install python3-dev python3-pip
apt-get install git   # <--  seems related to pybind11?
pip install zxing-cpp
```

**Comment by axxel** [comment, marked as answer]
> Can one pip-install without first installing git?

Yes, if a prebuilt wheel is available. That is generally the case for the last 3 python versions (at the time of release). right now this would be 3.9, 3.10 and 3.11.

I [updated](https://github.com/zxing-cpp/zxing-cpp/commit/ca96dab34a293055378fc795fbe27abafa45fb3f) the readme to mention git.

> apt-get install git   # <--  seems related to pybind11?

It is also required to get the c++ sources of zxing-cpp itself.

EDIT: Thanks to @stumpylog this is not the case for current HEAD anymore (and should show up in a future 2.1 release). The zxing-cpp sources are now included in the sdist package and pybind11 will be found on the system if it is available (but still downloaded via git if not).

**Comment by cbm755** [comment]
> It is also required to get the c++ sources of zxing-cpp itself.

That part doesn't make sense to me when installing a release...   Are the "c++ sources" not included in the .tar.gz from pypi?  viz, https://files.pythonhosted.org/packages/3d/19/a07c245a5603b5c84916066d782523697febbfe83daaee98e44b78813c24/zxing-cpp-2.0.0.tar.gz

Apologies for the naive questions, still learning the relationship between the Python library `zxing-cpp` and this (larger?) project.

**Comment by axxel** [comment]
> Are the "c++ sources" not included in the .tar.gz from pypi?

The (single) c++ source file of the python wrapper is included but not the core zxing-cpp library. The python module can be either build from the top-level cmake project where it would usually be dynamically linked to the c++ library or it can be built from the `setup.py` file as a 'standalone' (statically linked) python module. In the latter case, cmake determines whether the c++ sources are at `../../core`. If they are not, then we git-fetch them from github. Since the sources of pybind11 need to be fetched from there anyway, that does not make any practical difference.

**Reply by axxel** [reply]
This information is outdated since #530.

**Comment by stumpylog** [comment]
This setup prevents piwheels from building a wheel for armv7.  I think it would be an improvement to have the Python sdist include the library source files, making it a standalone source dist, instead of requiring Git.

**Reply by axxel** [reply]
Please note my comment above about pybind11, which is also needed and fetched via git (and will most certainly never be included in the sdist). So your issue would still be the same, right? The limitation you mentioned is only present for armv7 builds? how come?

**Reply by stumpylog** [reply]
PiWheels doesn't allow `git clone` when it tries to build, presumably for security reasons, though that's just a guess.  So the cloning done via CMake fails for both pybind and the actual library source.

Not super familiar with pybind, but wouldn't setting it as a requirement of the build system ensure it is available for doing a `find_package(pybind11 REQUIRED)` or something similar, without the clone.  Unless there's an unreleased feature that's needed, I suppose

**Reply by axxel** [reply]
I see your point. This is certainly doable but requires some change in how the source dist is created. Are you in a position to look into this? I'm currently prioritizing my time on QRCodes where some sponsoring is involved.

**Reply by stumpylog** [reply]
I'm more of a library user than packager, but I will try to poke around and can maybe get some initial working prototype
